File: 0204_count_primes/count_primes.go
package main

/**
 * Time: O(n)
 * Space: O(n)
 */
func countPrimes(n int) int {
	if n <= 2 {
		return 0
	}
	res := 1
	primes := make([]int, n)
	for i := 3; i < n; i += 2 {
		if primes[i] == 0 {
			res++
			for j := i * i; j < n; j += i {
				primes[j] = 1
			}
		}
	}
	return res
}

func main() {

}

File: 0204_count_primes/count_primes.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * Time: O(n) 
 * Space: O(n)
 */
int count_primes(int n)
{
    if (n <= 2)
        return 0;
    int res = 1, upper = sqrt(n);
    vector<int> primes(n, 0);
    for (int i = 3; i < n; i += 2)
        if (!primes[i])
        {
            res++;
            if (i > upper) // for overflow
                continue;
            for (int j = i * i; j < n; j += i)
                primes[j] = 1;
        }
    return res;
}
